### ISSUE TITLE
undefined symbol: __sycl_register_lib

### ISSUE THREAD

**Issue by jeffhammond** [intel/llvm#1110]
I installed 38ace4d50c3c61378197a4aea315c21063d28862 just now on my Hades Canyon NUC running Ubuntu 19.04 with all packages current.

```
$ clang++ -v
clang version 11.0.0 (https://github.com/intel/llvm 38ace4d50c3c61378197a4aea315c21063d28862)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/jrhammon/ISYCL/build/bin
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/8
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/9
Selected GCC installation: /usr/lib/gcc/x86_64-linux-gnu/8
Candidate multilib: .;@m64
Selected multilib: .;@m64
```

```
$ make -j sycl
clang++ -DPRKVERSION="2.16" -fsycl -fsycl-unnamed-lambda -std=c++17 -O3 -DUSE_SYCL -DUSE_2D_INDEXING=0 nstream-sycl.cc -o nstream-sycl
```

```
$ ./nstream-sycl 10 1000000
./nstream-sycl: symbol lookup error: ./nstream-sycl: undefined symbol: __sycl_register_lib
```

I don't know why this error occurs, but it shouldn't.

**Comment by bader** [comment]
Could you attach a source file, so I can try to reproduce this issue on my side, please?

**Comment by jeffhammond** [comment]
@bader `git clone https://github.com/ParRes/Kernels.git PRK && cd PRK/Cxx11`

**Comment by jeffhammond** [comment]
@jbrodman can't reproduce, so maybe my build recipe is wrong.
```sh
#!/bin/bash
export SYCL_HOME=/home/jrhammon/ISYCL
#cd $SYCL_HOME/llvm && time git checkout usmapi && time git pull
cd $SYCL_HOME/llvm && time git checkout sycl && time git pull
rm -rf $SYCL_HOME/build
mkdir -p $SYCL_HOME/build && \
    cd $SYCL_HOME/build && \
    time cmake \
        -DCMAKE_INSTALL_PREFIX=/opt/isycl \
        -DCMAKE_BUILD_TYPE=Release \
        -DLLVM_ENABLE_PROJECTS="clang;llvm-spirv;sycl" \
        -DLLVM_EXTERNAL_PROJECTS="llvm-spirv;sycl" \
        -DLLVM_EXTERNAL_SYCL_SOURCE_DIR=$SYCL_HOME/llvm/sycl \
        -DLLVM_EXTERNAL_LLVM_SPIRV_SOURCE_DIR=$SYCL_HOME/llvm/llvm-spirv \
        -DLLVM_TOOL_SYCL_BUILD=ON \
        -DLLVM_TOOL_LLVM_SPIRV_BUILD=ON \
        $SYCL_HOME/llvm/llvm && \
time make -j4 sycl-toolchain
time make -j4 sycl-toolchain install #DESTDIR=/opt/isycl
```

**Comment by jeffhammond** [comment]
This issue was caused by user error, although it is very frustrating that the toolchain does not check shared library ABI versioning or use rpath to ensure that the compiler builds applications that will use the right runtime library and not some other runtime library from last summer.

To be specific, I installed the toolchain into `/opt/isycl` but had an installation from June 2019 in `/usr/local` that I forgot about.  Because Linux looks for shared objects in `/usr/local`, it found `libsycl.so` there, rather than the one in `/opt/isycl/lib`.  It would be far better user experience if the compiler driver would add `-Wl,-rpath=$PREFIX/lib` so that users were not at the mercy of `LD_LIBRARY_PATH` and friends.
